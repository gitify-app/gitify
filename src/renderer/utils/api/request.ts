import axios, {
  type AxiosResponse,
  type AxiosPromise,
  type Method,
} from 'axios';
import log from 'electron-log';
import type { Link, Token } from '../../types';
import { getNextURLFromLinkHeader } from './utils';

/**
 * Perform an unauthenticated API request
 *
 * @param url
 * @param method
 * @param data
 * @returns
 */
export function apiRequest(
  url: Link,
  method: Method,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  return axios({ method, url, data });
}

/**
 * Perform an authenticated API request
 *
 * @param url
 * @param method
 * @param token
 * @param data
 * @param fetchAllRecords whether to fetch all records or just the first page
 * @returns
 */
export async function apiRequestAuth(
  url: Link,
  method: Method,
  token: Token,
  data = {},
  fetchAllRecords = false,
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common.Authorization = `token ${token}`;
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  axios.defaults.headers.common['Cache-Control'] = shouldRequestWithNoCache(url)
    ? 'no-cache'
    : '';

  if (!fetchAllRecords) {
    return axios({ method, url, data });
  }

  let response: AxiosResponse | null = null;
  let combinedData = [];

  try {
    let nextUrl: string | null = url;

    while (nextUrl) {
      response = await axios({ method, url: nextUrl, data });

      // If no data is returned, break the loop
      if (!response?.data) {
        break;
      }

      combinedData = combinedData.concat(response.data); // Accumulate data

      nextUrl = getNextURLFromLinkHeader(response);
    }
  } catch (error) {
    log.error('API request failed:', error);
    throw error;
  }

  return {
    ...response,
    data: combinedData,
  } as AxiosResponse;
}

/**
 * Return true if the request should be made with no-cache
 *
 * @param url
 * @returns boolean
 */
function shouldRequestWithNoCache(url: string) {
  const parsedUrl = new URL(url);

  switch (parsedUrl.pathname) {
    case '/notifications':
      return true;
    default:
      return false;
  }
}
