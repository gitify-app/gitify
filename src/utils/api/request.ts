import axios, { type AxiosPromise, type Method } from 'axios';
import type { Link, Token } from '../../types';

export function apiRequest(
  url: Link,
  method: Method,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  return axios({ method, url, data });
}

export function apiRequestAuth(
  url: Link,
  method: Method,
  token: Token,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common.Authorization = `token ${token}`;
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  return axios({ method, url, data });
}
