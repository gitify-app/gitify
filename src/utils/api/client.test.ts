import axios, { type AxiosPromise, type AxiosResponse } from 'axios';
import {
  mockGitHubCloudAccount,
  mockGitHubEnterpriseServerAccount,
  mockToken,
} from '../../__mocks__/state-mocks';
import type { Hostname, Link, SettingsState, Token } from '../../types';
import {
  getAuthenticatedUser,
  getHtmlUrl,
  headNotifications,
  ignoreNotificationThreadSubscription,
  listNotificationsForAuthenticatedUser,
  markNotificationThreadAsDone,
  markNotificationThreadAsRead,
  markRepositoryNotificationsAsRead,
} from './client';
import * as apiRequests from './request';

jest.mock('axios');

const mockGitHubHostname = 'github.com' as Hostname;
const mockEnterpriseHostname = 'example.com' as Hostname;
const mockThreadId = '1234';
const mockRepoSlug = 'gitify-app/notifications-test';

describe('utils/api/client.ts', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('getAuthenticatedUser', () => {
    it('should fetch authenticated user - github', async () => {
      await getAuthenticatedUser(mockGitHubHostname, mockToken);

      expect(axios).toHaveBeenCalledWith({
        url: 'https://api.github.com/user',
        method: 'GET',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should fetch authenticated user - enterprise', async () => {
      await getAuthenticatedUser(mockEnterpriseHostname, mockToken);

      expect(axios).toHaveBeenCalledWith({
        url: 'https://example.com/api/v3/user',
        method: 'GET',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('headNotifications', () => {
    it('should fetch notifications head - github', async () => {
      await headNotifications(mockGitHubHostname, mockToken);

      expect(axios).toHaveBeenCalledWith({
        url: 'https://api.github.com/notifications',
        method: 'HEAD',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should fetch notifications head - enterprise', async () => {
      await headNotifications(mockEnterpriseHostname, mockToken);

      expect(axios).toHaveBeenCalledWith({
        url: 'https://example.com/api/v3/notifications',
        method: 'HEAD',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('listNotificationsForAuthenticatedUser', () => {
    const mockSettings: Partial<SettingsState> = {
      participating: true,
    };

    it('should list notifications for user - github cloud', async () => {
      await listNotificationsForAuthenticatedUser(
        mockGitHubCloudAccount,
        mockSettings as SettingsState,
      );

      expect(axios).toHaveBeenCalledWith({
        url: 'https://api.github.com/notifications?participating=true',
        method: 'GET',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should list notifications for user - github enterprise server', async () => {
      await listNotificationsForAuthenticatedUser(
        mockGitHubEnterpriseServerAccount,
        mockSettings as SettingsState,
      );

      expect(axios).toHaveBeenCalledWith({
        url: 'https://github.gitify.io/api/v3/notifications?participating=true',
        method: 'GET',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('markNotificationThreadAsRead', () => {
    it('should mark notification thread as read - github', async () => {
      await markNotificationThreadAsRead(
        mockThreadId,
        mockGitHubHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://api.github.com/notifications/threads/${mockThreadId}`,
        method: 'PATCH',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should mark notification thread as read- enterprise', async () => {
      await markNotificationThreadAsRead(
        mockThreadId,
        mockEnterpriseHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://example.com/api/v3/notifications/threads/${mockThreadId}`,
        method: 'PATCH',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('markNotificationThreadAsDone', () => {
    it('should mark notification thread as done - github', async () => {
      await markNotificationThreadAsDone(
        mockThreadId,
        mockGitHubHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://api.github.com/notifications/threads/${mockThreadId}`,
        method: 'DELETE',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should mark notification thread as done- enterprise', async () => {
      await markNotificationThreadAsDone(
        mockThreadId,
        mockEnterpriseHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://example.com/api/v3/notifications/threads/${mockThreadId}`,
        method: 'DELETE',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('ignoreNotificationThreadSubscription', () => {
    it('should ignore notification thread subscription - github', async () => {
      await ignoreNotificationThreadSubscription(
        mockThreadId,
        mockGitHubHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://api.github.com/notifications/threads/${mockThreadId}/subscription`,
        method: 'PUT',
        data: { ignored: true },
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should ignore notification thread subscription- enterprise', async () => {
      await ignoreNotificationThreadSubscription(
        mockThreadId,
        mockEnterpriseHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://example.com/api/v3/notifications/threads/${mockThreadId}/subscription`,
        method: 'PUT',
        data: { ignored: true },
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('markRepositoryNotificationsAsRead', () => {
    it('should mark repository notifications as read - github', async () => {
      await markRepositoryNotificationsAsRead(
        mockRepoSlug,
        mockGitHubHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://api.github.com/repos/${mockRepoSlug}/notifications`,
        method: 'PUT',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });

    it('should mark repository notifications as read - enterprise', async () => {
      await markRepositoryNotificationsAsRead(
        mockRepoSlug,
        mockEnterpriseHostname,
        mockToken,
      );

      expect(axios).toHaveBeenCalledWith({
        url: `https://example.com/api/v3/repos/${mockRepoSlug}/notifications`,
        method: 'PUT',
        data: {},
      });

      expect(axios.defaults.headers.common).toMatchSnapshot();
    });
  });

  describe('getHtmlUrl', () => {
    it('should return the HTML URL', async () => {
      const apiRequestAuthMock = jest.spyOn(apiRequests, 'apiRequestAuth');

      const requestPromise = new Promise((resolve) =>
        resolve({
          data: {
            html_url:
              'https://github.com/gitify-app/notifications-test/issues/785',
          },
        } as AxiosResponse),
      ) as AxiosPromise;

      apiRequestAuthMock.mockResolvedValue(requestPromise);

      const result = await getHtmlUrl(
        'https://api.github.com/repos/gitify-app/notifications-test/issues/785' as Link,
        '123' as Token,
      );
      expect(result).toBe(
        'https://github.com/gitify-app/notifications-test/issues/785',
      );
    });

    it('should handle error', async () => {
      const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();

      const apiRequestAuthMock = jest.spyOn(apiRequests, 'apiRequestAuth');

      const mockError = new Error('Test error');

      apiRequestAuthMock.mockRejectedValue(mockError);

      await getHtmlUrl(
        'https://api.github.com/repos/gitify-app/gitify/issues/785' as Link,
        '123' as Token,
      );

      expect(consoleErrorSpy).toHaveBeenCalledWith('Failed to get html url');
    });
  });
});
