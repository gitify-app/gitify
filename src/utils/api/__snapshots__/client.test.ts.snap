// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/api/client.ts getAuthenticatedUser should fetch authenticated user - enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts getAuthenticatedUser should fetch authenticated user - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts getRootHypermediaLinks should fetch root hypermedia links - enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts getRootHypermediaLinks should fetch root hypermedia links - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts headNotifications should fetch notifications head - enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts headNotifications should fetch notifications head - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts ignoreNotificationThreadSubscription should ignore notification thread subscription - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts ignoreNotificationThreadSubscription should ignore notification thread subscription- enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts listNotificationsForAuthenticatedUser should list notifications for user - enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts listNotificationsForAuthenticatedUser should list notifications for user - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts markNotificationThreadAsDone should mark notification thread as done - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts markNotificationThreadAsDone should mark notification thread as done- enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts markNotificationThreadAsRead should mark notification thread as read - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts markNotificationThreadAsRead should mark notification thread as read- enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts markRepositoryNotificationsAsRead should mark repository notifications as read - enterprise 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/client.ts markRepositoryNotificationsAsRead should mark repository notifications as read - github 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;
